File: packages/igloo-ts/src/widgets/svg.ts
import { Svg as WitSvg } from 'iced:app/svg@0.1.0';
import { Element, type IntoElement } from '../element.js';

/**
 * Builder for creating SVG widgets.
 * An SVG displays a vector graphics image from a file path.
 *
 * @example
 * ```typescript
 * const icon = Svg.new('/path/to/icon.svg');
 * ```
 */
export class Svg implements IntoElement<never> {
  private raw: WitSvg;

  private constructor(path: string) {
    this.raw = new WitSvg(path);
  }

  /** Create a new SVG builder with the given file path */
  static new(path: string): Svg {
    return new Svg(path);
  }

  /** Convert to Element */
  intoElement(): Element<never> {
    return new Element(WitSvg.intoElement(this.raw));
  }
}

File: crates/igloo_guest/src/widgets/svg.rs
use crate::Element;
use crate::bindings::iced::app::widgets::{Node, SvgNode};

/// A vector graphics image.
pub struct Svg {
    path: String,
}

impl Svg {
    /// Creates a new [`Svg`] from the given path.
    pub fn new(path: impl Into<String>) -> Self {
        Self { path: path.into() }
    }
}

impl<Message: 'static> From<Svg> for Element<Message> {
    fn from(svg: Svg) -> Self {
        Element::new(move |_realize, arena| arena.push(Node::Svg(SvgNode { path: svg.path })))
    }
}
